src/langchain_google_spanner/graph_qa.py

```diff
@@ -0,0 +1,375 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import re
+from typing import Any, Dict, List, Optional
+
+from langchain.chains.base import Chain
+from langchain_core.callbacks import CallbackManagerForChainRun
+from langchain_core.language_models import BaseLanguageModel
+from langchain_core.output_parsers import JsonOutputParser, StrOutputParser
+from langchain_core.prompts import BasePromptTemplate
+from langchain_core.prompts.prompt import PromptTemplate
+from langchain_core.runnables import RunnableSequence
+from pydantic.v1 import BaseModel, Field
+
+from langchain_google_spanner.graph_store import SpannerGraphStore
+
+from .prompts import (
+    DEFAULT_GQL_FIX_TEMPLATE,
+    DEFAULT_GQL_TEMPLATE,
+    DEFAULT_GQL_VERIFY_TEMPLATE,
+    SPANNERGRAPH_QA_TEMPLATE,
+)
+
+GQL_GENERATION_PROMPT = PromptTemplate(
+    template=DEFAULT_GQL_TEMPLATE,
+    input_variables=["question", "schema"],
+)
+
+
+class VerifyGqlOutput(BaseModel):
+    input_gql: str
+    made_change: bool
+    explanation: str
+    verified_gql: str
+
+
+verify_gql_output_parser = JsonOutputParser(pydantic_object=VerifyGqlOutput)
+
+GQL_VERIFY_PROMPT = PromptTemplate(
+    template=DEFAULT_GQL_VERIFY_TEMPLATE,
+    input_variables=["question", "generated_gql", "graph_schema"],
+    partial_variables={
+        "format_instructions": verify_gql_output_parser.get_format_instructions()
+    },
+)
+
+GQL_FIX_PROMPT = PromptTemplate(
+    template=DEFAULT_GQL_FIX_TEMPLATE,
+    input_variables=["question", "generated_gql", "err_msg", "schema"],
+)
+
+SPANNERGRAPH_QA_PROMPT = PromptTemplate(
+    template=SPANNERGRAPH_QA_TEMPLATE,
+    input_variables=["question", "graph_schema", "graph_query", "context"],
+)
+
+INTERMEDIATE_STEPS_KEY = "intermediate_steps"
+
+
+def fix_gql_syntax(query: str) -> str:
+    """Fixes the syntax of a GQL query.
+
+    Args:
+        query: The input GQL query.
+
+    Returns:
+        Possibly modified GQL query.
+    """
+
+    query = re.sub(r"-\[(.*?):(\w+)\*(\d+)\.\.(\d+)\]->", r"-[\1:\2]->{\3,\4}", query)
```

Comment:
nit: could you add a few examples so that it's easier to understand.

Reply:
done